### ISSUE TITLE
[MNG-4706] Multithreaded building can create bad files for downloaded artifacts in local repository

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6374]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** opened **[MNG-4706](https://issues.apache.org/jira/browse/MNG-4706?redirect=false)** and commented

Within a parallel build, modules A and B can start at the same time and decide at the same time that they need to download artifact X. Since X is not available locally, both can start download of X. Both will download X and both will try to write the file for X to disk.

On Windows platforms, one of them can fail with file locking problems. On Unixes the contents of the written file can be non-deterministic.

This issue is related to MNG-2802 in the sense that it can create a corrupted artifact file in the local repository, and it can happen to ANY file written. But as opposed to MNG-2802 the synhronization needed to fix this issue is basically only within a single running maven process, so regular java mechanisms can apply, whereas MNG-2802 probably needs to use a totally different solution space.

---

**Affects:** 3.0-beta-1

**Issue Links:**
- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository
 (_**"is a child of"**_)
- [MNG-7545](https://issues.apache.org/jira/browse/MNG-7545) Multi building can create bad files for downloaded artifacts in local repository
 (_**"is cloned by"**_)
- [MNG-4742](https://issues.apache.org/jira/browse/MNG-4742) Multi-threaded artifact resolver corrupts artifacts on download
 (_**"is duplicated by"**_)
- [MRESOLVER-134](https://issues.apache.org/jira/browse/MRESOLVER-134) Unsatisfiable Range header causing 416 HTTP error
 (_**"causes"**_)
- [MNG-6604](https://issues.apache.org/jira/browse/MNG-6604) Intermittent failures while downloading GAVs from Nexus

- [MNG-5315](https://issues.apache.org/jira/browse/MNG-5315) Artifact resolution sporadically fails in parallel builds

- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository

- [MNG-6996](https://issues.apache.org/jira/browse/MNG-6996) Upgrade Maven Resolver to 1.6.1
 (_**"requires"**_)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

18 votes, 21 watchers

**Comment by jira-importer** [comment]
**[Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf)** commented

Adding the following two lines to .mvn/maven.config solves the problem for me, or at least makes it much much less likely to occur with latest maven 3.3.x builds (circa 3.3.6)

```
-Daether.connector.basic.threads=1
-Daether.connector.resumeDownloads=false
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

MRESOLVER-131 introduces a fully new, multi-JVM concurrent approach based on Redis. Try it from here: https://github.com/apache/maven-resolver/tree/MRESOLVER-131. Read it from here: https://maven.apache.org/resolver-archives/resolver-LATEST/maven-resolver-synccontext-redisson/index.html

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Use any of the two new sync contexts to solve this concurrency issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Maven Resolver 1.6.2 will be available with Maven 3.8.2 so this has been addressed in 3.8.x too.

**Comment by jira-importer** [comment]
**[liuchangcheng](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER296009)** commented

is this problem fixed? we are still facing this problem with maven version 3.8.3
